torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -482,6 +488,43 @@ ProcessGroupNCCL::ProcessGroupNCCL(
     asyncErrorHandling_ = false;
   }
 
+  // Perform health check by initializing dummy communicators and destroying
+  // them. This will help indicate any NCCL-related issues prior to the first
+  // collective.
+  // Run it in a separate thread and wait on CV to handle timeouts, since
+  // majority of getNCCLComm failures are hangs.
+  std::mutex healthCheckMutex;
```

Comment:
nit: can use atomic bool for healthCheckSuccess and no need for this mutex then?

Reply:
I think we still need the lock due to the condition variable requiring a lock? 

Also I read this article: https://www.modernescpp.com/index.php/c-core-guidelines-be-aware-of-the-traps-of-condition-variables that said an atomic var is not enough to avoid races in some case. Most of the logic around signaling the main thread comes from there, does it seem okay?